### PR TITLE
Rego profiler

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
Changes to add a Rego expression profiler. 

Sample output for an expression:
`{
            "index": 0,
            "totalTimeNs": 72225,
            "numEval": 4,
            "numRedo": 4,
            "row": 1,
            "col": 1,
            "text": "x = 1"
          }
     `

### CODE REVIEW

File: profiler/profiler.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+// Profiler computes and reports on the time spent on expressions
+type Profiler struct {
+	hits        map[string]map[Position]ExprStats
+	activeTimer time.Time
+	prevExpr    ExprInfo
+}
+
+// ExprInfo stores information about an expression
+type ExprInfo struct {
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply profiler/profiler.go]: Unexported `ExprInfo`

File: profiler/profiler.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	// record the profiler results for the previous expression
+	hits, ok := p.hits[p.prevExpr.file]
+	if !ok {
+		hits = map[Position]ExprStats{}
+		hits[Position{p.prevExpr.row}] = getProfilerStats(p.prevExpr, p.activeTimer)
+		p.hits[p.prevExpr.file] = hits
+	} else {
+		pStats, ok := hits[Position{p.prevExpr.row}]
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply profiler/profiler.go]: Done

File: profiler/profiler.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	} else {
+		profilerStats.NumRedo = 1
+	}
+
+	return profilerStats
+}
+
+// ExprStats represents the result of profiling an expression
+type ExprStats struct {
+	Index       int    `json:"index"`
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply profiler/profiler.go]: Removed index

File: profiler/profiler.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	TotalTimeNs int64  `json:"totalTimeNs"`
+	NumEval     int    `json:"numEval"`
+	NumRedo     int    `json:"numRedo"`
+	Row         int    `json:"row"`
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply profiler/profiler.go]: Done

File: profiler/profiler_test.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	if !ok {
+		t.Fatal("Expected file report for test.rego")
+	}
+
+	if len(fr.Result) != 11 {
+		t.Fatalf("Expected file report length to be 11 instead got %v", len(fr.Result))
+	}
+}
+
+func TestProfileSleepCmd(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply profiler/profiler_test.go]: changed name to `TestProfileCheckExprDuration`

[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply profiler/profiler_test.go]: Updated the tests

File: profiler/profiler_test.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	topdown.RegisterFunctionalBuiltin1("test.sleep", func(a ast.Value) (ast.Value, error) {
+		d, _ := time.ParseDuration(string(a.(ast.String)))
+		time.Sleep(d)
+		return ast.Null{}, nil
+	})
+
+	module := `package test
+
+	foo { 
+	  test.sleep("1s")
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply profiler/profiler_test.go]: Done

File: profiler/profiler.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+		profilerStats.NumRedo = 1
+	}
+
+	return profilerStats
+}
+
+// ExprStats represents the result of profiling an expression
+type ExprStats struct {
+	Index       int    `json:"index"`
+	TotalTimeNs int64  `json:"totalTimeNs"`
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply profiler/profiler.go]: Done

File: profiler/profiler.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+			p.processExpr(expr, "Eval")
+		}
+	case topdown.RedoOp:
+		if expr, ok := event.Node.(*ast.Expr); ok && expr != nil {
+			p.processExpr(expr, "Redo")
+		}
+	}
+}
+
+func (p *Profiler) processExpr(expr *ast.Expr, op string) {
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply profiler/profiler.go]: Done

File: profiler/profiler.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+type Profiler struct {
+	hits        map[string]map[Position]ExprStats
+	activeTimer time.Time
+	prevExpr    ExprInfo
+}
+
+// ExprInfo stores information about an expression
+type ExprInfo struct {
+	index int
+	file  string
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply profiler/profiler.go]: Done

File: profiler/profiler.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
+// Copyright 2018 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+// Package profiler computes and reports on the time spent on expressions
```
[the remark this answers is not in the document]
Comment (ashutosh-narkar) [reply profiler/profiler.go]: Done

File: profiler/profiler.go [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+func (p *Profiler) processLastExpr() {
+	expr := ast.Expr{
+		Location: p.prevExpr.location,
+	}
+	p.processExpr(&expr, p.prevExpr.op)
+}
+
+// Position represents a file location.
+type Position struct {
```
Comment (tsandall) [inline profiler/profiler.go]: It looks like this doesn't have to be exported. I would replace references to it simply with an `int`.

File: profiler/profiler_test.go [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+		t.Fatalf("Expected file report length to be 11 instead got %v", len(fr.Result))
+	}
+
+	expectedNumEval := []int{1, 2, 1, 1, 1, 1633, 1, 1, 1, 1, 1}
+	expectedNumRedo := []int{1, 0, 0, 1, 1633, 0, 1, 1, 1, 1, 0}
+	expectedRow := []int{4, 5, 6, 10, 11, 12, 16, 17, 18, 22, 23}
+
+	for idx, actualExprStat := range fr.Result {
+		if actualExprStat.NumEval != expectedNumEval[idx] {
+			t.Fatalf("Expected number of evals %v but got %v", expectedNumEval[idx], actualExprStat.NumEval)
```
Comment (tsandall) [inline profiler/profiler_test.go]: It would be helpful to have the `idx` value in the test failures here.

**Pull request merged**